src/fa.c

```diff
@@ -2217,8 +2227,12 @@ struct fa *fa_intersect(struct fa *fa1, struct fa *fa2) {
         return fa_make_empty();
 
     if (fa1->nocase != fa2->nocase) {
-        F(case_expand(fa1));
-        F(case_expand(fa2));
+        cfa1 = fa_clone(fa1);
+        cfa2 = fa_clone(fa2);
+        if (cfa1 == NULL || cfa2 == NULL)
+            goto error;
+        F(case_expand(cfa1));
+        F(case_expand(cfa2));
```

Comment:
Couldn't we now do `fa1 = cfa1` so that we can avoid all the places where we have `(cfa1 ? cfa1 : fa1)` ?